Initial Governance contract integration

AnnaShaleva opened the pull request:
Here's a draft that I have right now, several questions need to be decided, they are marked as TODOs in the code. Close #14.

~~The good news is that the code is able to call Governance API and get the current set of validators. The bad news is that sometimes contract call doesn't work as expected, the following error occurs on attempt to unpack the result, I'm investigating it.~~
UPD: the described problem is solved, now Governance calls work as expected. Now I'm working on the rest of less critical TODOs:
- [x] Decide about final contract hash (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519684969).
- [x] ~~Automate contract build and genesis allocations pre-filling.~~ Will be done in #149.
- [x] Make contract code network-independent (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519655264 and https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519662507) and update the rest of privnet/testnet configurations afterwards.
- [x] Deal with genesis-based contract call processing.
- [x] Deal with next epoch validators calculations (ref. https://github.com/bane-labs/go-ethereum/pull/146#discussion_r1519970256).
- [x] Add validators cache.

AnnaShaleva committed:
dbft: do not call postBlock callback immediately in ProcessBlock

There's a chance that newly-accepted block will be added by worker using
the corresponding sealing task saved in the worker state. If so, then
BlockQueue don't have any way to track the end of block addition
process. Since postBlock contains NextConsensus computations, it must be
called only after successfull block persist so that the proper state is
available in the chain. And thus, call postBlock only when chain
notification arrives.

A part of #14. An example of problem logs that this commit is aimed to
fix:
```
2024-03-11T19:52:22.324+0300	DEBUG	dbft@v0.0.0-20240130131324-6d59675ad370/dbft.go:214	received message	{"type": "Commit", "from": 5, "height": 1, "view": 1, "my_height": 1, "my_view": 1}
2024-03-11T19:52:22.324+0300	DEBUG	dbft@v0.0.0-20240130131324-6d59675ad370/dbft.go:214	received message	{"type": "Commit", "from": 6, "height": 1, "view": 1, "my_height": 1, "my_view": 1}
2024-03-11T19:52:22.324+0300	INFO	dbft@v0.0.0-20240130131324-6d59675ad370/dbft.go:511	received Commit	{"validator": 6}
2024-03-11T19:52:22.325+0300	INFO	dbft@v0.0.0-20240130131324-6d59675ad370/check.go:66	approving block	{"height": 1, "hash": "8633939a9fa52e0fc260b8dc335f21bf11c5519ea150683d434e4986f597a294", "tx_count": 0, "merkle": "eeccdc5021307b4b06231fcd6c7b7a9c632863d647459bb4124e9c95f26e50f5", "prev": "9ee34b625f2e9aac3d0b7aaaea40db845906d6d24c89e3e4f53af8d63074eb83"}
INFO [03-11|19:52:22.325] calculating validators                   height=1
CRIT [03-11|19:52:22.325] failed to compute next block validators  err="header not found"
```

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: use Governance contract for validators calculations

A part of #14. Refer to Governance contract to get next validators
addresses.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@chenquanyu, @txhsl welcome to review, but please keep in mind that it's a very early version of integration.

@nicolegys, this PR is not ready for testing yet, I'll notify you as soon as it's ready.

AnnaShaleva committed:
dbft: make payload verification error verbose

For better further debugging experience.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: retrieve SB validators from Governance contract at genesis

Genesis-based validators retrieval may be done using the same approach
as for further blocks. This change allows us to remove StandByValidators
from dBFT configuration in future.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: always recalculate block state got from miner

We need an access to block state at the PrepareRequest construction
level to be able to calculate NextConsensus for the current proposal.
State for the 0-th view can be fetched from miner, but I'd avoid adding
one more miner-dependent state field to the dBFT engine, especially
given the fact that one day we'll remove state calculations from
miner.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: change the scheme of NextConsensus calculation

Calculate NextConsensus for accepting block using the state of
accepting block. The state is retrieved either during PrepareRequest
construction (by Primary) or during PrepareRequest verification (by
backups). Thanks to this change no additional APIs are required from
Governance contract, the only `getCurrentConsensus` API is necessary.

A part of #14.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: cache validators computed based on the accepted state

If validators are requested by block height then the block state is
already in chain and accepted, and thus, the set of validators
computed based on this state won't change.

Cache validators for three most recent heights which is enough for
effective network payloads processing (we may have stale payloads
travelling through the network) and dBFT fuctioning (it always needs
the latest set of validators for block signing).

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: rename getNextBlockValidators to getValidators

This method returns exactly validators of the latest finalized epoch
from the specified block/state PoW.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@roman-khimov, first version is ready for review. The contract version should be updated (currently the very old one is used to test 7-nodes privnet), but the idea of the PR won't be changed if no issues found.

chenquanyu commented on consensus/dbft/dbft.go:
```diff
@@ -1589,56 +1610,81 @@ func (c *DBFT) APIs(chain consensus.ChainHeaderReader) []rpc.API {
+
+	// TODO: roman-khimov, we've agreed that validators are sorted using lexicographic order, but in contract they are
+	// sorted by votes. Need to come to some single solution.
+	slices.SortFunc(res, common.Address.Cmp)
```
Do we need to sort validators in contract? It seems to be enough to sort them here.

AnnaShaleva replied:
> Do we need to sort validators in contract?

We don't. But I'd like to ensure, that it's OK for us to sort validators on dBFT engine side.

AnnaShaleva replied:
It's OK for us, so let this sorting code to be present here. TODO is removed.

chenquanyu commented on core/systemcontracts/contracts.go:
```diff
@@ -2,7 +2,10 @@ package systemcontracts
+	// TODO: need to finalize genesis contract hashes in https://github.com/bane-labs/go-ethereum/pull/146 and https://github.com/bane-labs/go-ethereum/pull/7
+	GovernanceHash = "0x1212000000000000000000000000000000000001"
+	// GovernanceABI is a compiled ABI of Governance contract.
+	GovernanceABI = `[{"anonymous":false,"inputs":[{"indexed":false,"internalType":"address","name":"candidate","type":"address"},{"indexed":false,"internalType":"uint256","name":"reward","type":"uint256"}],"name":"CandidateClaim","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"internalType":"address","name":"candidate","type":"address"}],"name":"Exit","type":"event"},{"anonymous":fal ...
```
May be we only need the interface abi here.

AnnaShaleva replied:
Yes, it will be updated once we have Governance finalized.

AnnaShaleva committed:
systemcontracts: decode system contracts hashes/ABI in-place

No errors expected during decoding since these are trusted data.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: implement OnPersist system call

Use the same approach as for Beacon block root processing via system
contract.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: do not finalize block twise on PrepareRequest verification

Finalize is called by state processor during PrepareRequest state
verification, and thus, don't need to call it one more time manually
in WithNewPrepareRequest callback.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
*: update 7-nodes privnet configuration

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

chenquanyu commented on core/state_processor.go:
```diff
@@ -189,3 +196,28 @@ func ProcessBeaconBlockRoot(beaconRoot common.Hash, vmenv *vm.EVM, statedb *stat
 }
+
+// ProcessOnPersist applies a system call to the governance contract.
+func ProcessOnPersist(vmenv *vm.EVM, statedb *state.StateDB) error {
```
Now we perform a call to OnPerist method of Governance contract via direct state modification (without fully-qualified transaction creation; for free) before in-block transactions processing. I think we can raise an issue first and we can implement free transactions in the future.

AnnaShaleva replied:
Ref. #152.

AnnaShaleva commented:
@roman-khimov, ready for review & merge, @nicolegys is performing 7-nodes privnet testing. Should we wait for test results before the merge?

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -1593,7 +1601,8 @@ func (c *DBFT) APIs(chain consensus.ChainHeaderReader) []rpc.API {
 func (c *DBFT) getNextBlockValidators(blockHash common.Hash, blockNum uint64, compute bool) ([]common.Address, error) {
 	// Currently we don't have governance contract, thus, always return standby set.
-	if true {
+	// TODO: to be removed.
```
TODOs in the final commit series?

AnnaShaleva replied:
These TODOs are resolved in the subsequent commits. Fixed.

Just FYI, we have a set of similar commits in NeoGo :D

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -1603,26 +1612,33 @@ func (c *DBFT) getNextBlockValidators(blockHash common.Hash, blockNum uint64, co
+		// is `getNextConsensus` that returns current epoch validators for the middle of dBFT epoch and next epoch validators
+		// for the last block in the epoch. We can organize this at the code-level, but it complicates getNextBlockValidators call, so
+		// why not to move it to the contract level where it costs nothing.
+		// @roman-khimov, agree?
```
TODOs in the final commit series?

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -1603,26 +1612,33 @@ func (c *DBFT) getNextBlockValidators(blockHash common.Hash, blockNum uint64, co
 
 	ctx, cancel := context.WithCancel(context.Background())
-	// Cancel when we are finished consuming integers
+	// Cancel when we are finished consuming integers.
```
Mixed with other changes.

AnnaShaleva replied:
Moved to a separate commit.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -1632,9 +1648,19 @@ func (c *DBFT) getNextBlockValidators(blockHash common.Hash, blockNum uint64, co
 	}
 
-	var valSet []common.Address
+	var valSet [7]common.Address
```
Can we get this number from the contract?

AnnaShaleva replied:
It was a kind of an intermediate contract version, it's fixed in the later commits. But now when we have a relatively stable contract, we don't need a code for intermediate version, so removed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -455,9 +451,16 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
 				log.Crit("failed to finalize and assemble proposal",
 					"err", err)
 			}
-
```
Whitespace.

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -573,6 +573,27 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
+			// TODO: technically, the contract itself doesn't change the state in getCurrentValidators
+			// (at least, at the current version), and we don't need to create a copy. But I'm not sure
+			// whether processing engine makes changes in the state, it should be investigated in a separate
+			// issue. And if not, then state may not be copied.
```
Number?

AnnaShaleva replied:
[Number](https://github.com/bane-labs/go-ethereum/issues/154).

roman-khimov commented on internal/ethapi/api.go:
```diff
@@ -1138,6 +1138,12 @@ func DoCall(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash
 	return doCall(ctx, b, args, state, header, overrides, blockOverrides, timeout, globalGasCap)
 }
 
+func DoCallAtState(ctx context.Context, b Backend, args TransactionArgs, state *state.StateDB, header *types.Header, overrides *StateOverride, blockOverrides *BlockOverrides, timeout time.Duration, globalGasCap uint64) (*core.ExecutionResult, error) {
```
Undocumented?

AnnaShaleva replied:
I followed the existing `DoCall` style, it also doesn't have a comment. But probably we may write better code. Fixed.

roman-khimov commented on core/systemcontracts/contracts.go:
```diff
@@ -1,11 +1,18 @@
+	GovernanceHash = common.HexToAddress("0x1212000000000000000000000000000000000001")
 	// GovernanceABI is a compiled ABI of Governance contract.
-	GovernanceABI = `[{"anonymous":false,"inputs":[{"indexed":false,"internalType":"address","name":"candidate","type":"address"},{"indexed":false,"internalType":"uint256","name":"reward","type":"uint256"}],"name":"CandidateClaim","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"internalType":"address","name":"candidate","type":"address"}],"name":"Exit","type":"event"},{"anonymous":fal ...
+	GovernanceABI, _ = abi.JSON(strings.NewReader(`[{"anonymous":false,"inputs":[{"indexed":false,"internalType":"address","name":"candidate","type":"address"},{"indexed":false,"internalType":"uint256","name":"amount","type":"uint256"}],"name":"CandidateWithdraw","type":"event"},{"anonymous":false,"inputs":[{"indexed":false,"internalType":"address","name":"candidate","type":"address"}],"name":"Exit" ...
```
You can do this in `init` and panic.

AnnaShaleva replied:
Fixed.

roman-khimov commented:
We need to test voting that changes the set of validators, btw.

nicolegys commented:
CNs failed when trying to change epoch.

```
CRIT [03-27|08:16:07.046] failed to process state from proposal    err="failed to process block: could not apply OnPersist [0x1774aa018de852e473c1d5cf332fc22ebfe7af960b8d7eb64370e9fa20c4354d]: onPersist call failed: execution reverted" number=12 "seal hash"=c98122..e91b71 "parent hash"=0x40f5b34ede642c91ea99563ef4f0bbe7d384604fa4c6d2db4c39df4e2e7bc517 "intermediate merkle root"=047e7d..96635e coinbase=0x1212000000000000000000000000000000000003 "gas limit"=30,000,000 "gas used"=0 difficulty=1 "mix digest"=0x0000000000000000000000000000000000000000000000000000000000000000 nonce=3 time=1,711,527,227 "uncle hash"=0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347 txs=0
```

txhsl commented:
> CNs failed when trying to change epoch.

Have you tried to register at least 7 different addresses as candidates before epoch change? The related contract method is https://github.com/bane-labs/go-ethereum/blob/governance/contracts/solidity/Governance.sol#L119.

AnnaShaleva commented:
> CNs failed when trying to change epoch.

Node expects the contract to properly handle `onPersist` call, and if it's not the case, then the node have to stop operation. The case is that it's Primary node who is trying to construct PrepareRequest, and if the construction fails, then we're toast.

> Have you tried to register at least 7 different addresses as candidates before epoch change?

So I'd expect the contract to handle the case when not enough candidates are registered. In this case StandBy validators should be returned. Because we will have such cases in the real network.

AnnaShaleva committed:
dbft: fix comment format

No functional changes.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>
